### DISCUSSION TITLE
Are the segments of JUnit Jupiter unique IDs part of the public API?

### DISCUSSION THREAD

**Discussion by fmeum** [junit-team/junit-framework#3551] [Q&A] [answered]
I couldn't find any documentation on whether the precise format of JUnit Jupiter test "unique IDs" is guaranteed to be stable. In particular, can it be relied on that e.g. test template method invocations have `test-template` and `invocation` segments?

**Comment by sbrannen** [comment, marked as answer]
If you are referring to the `String` form of a unique ID, then no, that format is not guaranteed to be stable.

Although it's not very likely that the segment names will change drastically, they could still change if a technical need arises.

So I'd be hesitant to introduce any code/parsing that relies on the current format.

If you need to work with segments, you should use the `org.junit.platform.engine.UniqueId` API.

I believe the `"engine"` segment type is the only named element that is documented as stable.

@marcphilipp, please correct me if I'm recalling this incorrectly.

**Reply by fmeum** [reply]
Thanks for the information!

I am interested in the semantic form of a `UniqueId`, e.g., which segments it has, not the exact way these segments are serialized to a `String`, as the latter is taken care of by the `UniqueId` API as you mentioned.

I am asking because I don't see a way to reason about test structures given just an `ExtensionContext`. For example, it doesn't seem to be possible to tell whether an extension method is called for a test template or for which invocation of such a template. A natural approach would be to look at the `UniqueId` structure, but if that isn't public API, that's not a great way to go about it.
